importlib_metadata/__init__.py

```diff
@@ -151,6 +151,47 @@ def __getitem__(self, item):
         return self._key()[item]
 
 
+class Deps(set):
+    """
+    An iterable of packaging.requirement.Requirements.
+    """
+
+    @property
+    def missing(self):
+        """
+        Resolve which requirements are missing from the
+        current environment. Requires 'packaging' to be
+        installed.
+        """
+        return itertools.filterfalse(self._is_installed, self)
+
+    @staticmethod
+    def _is_installed(req):
+        from packaging.version import Version
+
+        try:
+            dist = distribution(req.name)
+        except PackageNotFoundError:
+            return False
+        found_ver = Version(dist.version)
+        return found_ver in req.specifier
+
+    @classmethod
+    def for_entry_point(cls, ep):
+        """
+        Return Deps for any extras associated with the entry point.
+
+        Requires ep.dist to be defined.
+        """
+        from packaging.requirements import Requirement
+        return cls(
+            req
+            for req in map(Requirement, always_iterable(ep.dist.requires))
+            for extra in ep.extras
+            if req.marker.evaluate(dict(extra=extra))
```

Comment:
```suggestion
            for extra in [None] + ep.extras
            if  req.marker is None or req.marker.evaluate(dict(extra=extra))
```

Reply:
It looks like there may be two untested use-cases you're attempting to address here.

1. req.marker could be None, so would trigger an AttributeError.
2. The null extra should be implied.

I think I agree with (1) and I'll try to expand the tests to capture that expectation.

For (2), I'm less sure - under what conditions would an entry point need the null extra? Can this project rely on the expectation that if a package is installed that its non-extra dependencies are installed?

Reply:
I've addressed (1) in 7e07b739ff84.

Reply:
It's not just the case of a missing dependency, what about if the dependency is there, but not compatible?

IMHO, if the feature does not handle all those use cases (missing / conflicting non-extra dependencies, or transitive dependencies), then what's the point? Better to not support it at all and be up-front about it in the documentation.

But considering the fact that an installed package metadata does not track requested extras (which means that something like `pip check` can't possibly detect a number of issues), accurately checking for missing / conflicting requirements when loading an entry point is an important feature. Even if apparently PYPA has decided to forego any kind of sanity check in generated console scripts launcher...

Reply:
Acknowledged. Thanks for explaining.

I do believe what you're seeking is a feature that I'd like to consider separately - that of verifying that a Distribution and its transitive dependencies are satisfied.

This feature is meant primarily to determine which additional dependencies are required for a given entry point and to report any that aren't already satisfied, leaving it to another function to handle what to do about that condition (install the dependency, raise an error, log a warning, ...).

I'd like to leave it as an additional exercise/feature to perform checks.

But now that I think about it, perhaps your right and this feature adds little value. After all, whatever feature checks for the required dependencies of an entry point may still want to check that the dependencies are met for the distribution (with relevant extras).

Something like `iter_unmet_dependencies(dist, extras=[])`, which could be called with `ep.dist, ep.extras` to validate an entry point.

At this point, I'm convinced the proposed feature and the 'unmet dependencies' feature is undesirable for this package (mainly because 'packaging' is unavailable). I'll work on creating a separate library that provides this functionality.

Reply:
BTW, I did try running the tests with:

```diff
diff --git a/importlib_metadata/__init__.py b/importlib_metadata/__init__.py
index a9df4ccf4846..346f81c70149 100644
--- a/importlib_metadata/__init__.py
+++ b/importlib_metadata/__init__.py
@@ -188,7 +188,7 @@ class Deps(set):
         return cls(
             req
             for req in map(Requirement, always_iterable(ep.dist.requires))
-            for extra in ep.extras
+            for extra in [None] + ep.extras
             if req.marker and req.marker.evaluate(dict(extra=extra))
         )
 ```

And the tests still pass, indicating that the technique of a `None` extra doesn't match the null marker. It probably would need to include `if not extra or (req.marker and req.marker.evaluate(...))`.

Reply:
I don't understand. If you also want the non-optional dependencies, then you need to change the test to `req.marker is None or req.marker.evaluate(dict(extra=extra))`. `not extra or (req.marker and req.marker.evaluate(...))` would mean that a requirement of the form `python-xlib; "linux" in sys_platform` is not ignored on non-Linux platforms, no?

Reply:
Yes, good point. In a subsequent implementation (nspektr), I separate the concerns of resolving extra dependencies and regular dependencies and use a NullMarker to resolve an object on which `.evaluate` is always run.